pkg/controller/installation/windows.go

```diff
@@ -0,0 +1,648 @@
+// run. In addition to any error, this function returns true if the upgrade
+// status has changed.
+func (w *calicoWindowsUpgrader) syncNodesToUpgrade() (bool, error) {
+	currNodesToUpgrade := make(map[string]*corev1.Node)
+	currNodesFinishedUpgrade := make(map[string]*corev1.Node)
+	currNodesUpgrading := make(map[string]*corev1.Node)
+
+	for _, obj := range w.nodeIndexer.List() {
+		node, ok := obj.(*corev1.Node)
+		if !ok {
+			return false, fmt.Errorf("Never expected index to have anything other than a Node object: %v", obj)
+		}
+
+		if node.Labels[corev1.LabelOSStable] != "windows" {
+			continue
+		}
+
+		windowsLog.V(1).Info(fmt.Sprintf("Processing node %v", node.Name))
+		exists, version := common.GetWindowsNodeVersion(node)
+		// If the version annotation doesn't exist, something is wrong with the
+		// Calico Windows node or it needs to be upgraded manually to
+		// a version supported by the calicoWindowsUpgrader.
+		if !exists {
+			return false, fmt.Errorf("Node %v does not have the version annotation, it might be unhealthy or it might be running an unsupported Calico version.", node.Name)
+		}
+
+		if version != w.expectedVersion {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v doesn't have the latest version. version=%v, expectedVersion=%v", node.Name, version, w.expectedVersion))
+			// If the version is outdated and the node does not yet have the upgrade
+			// label, we need to add the label to trigger the upgrade.
+			if _, exists := node.Labels[common.CalicoWindowsUpgradeScriptLabel]; !exists {
+				windowsLog.V(1).Info(fmt.Sprintf("Node %v doesn't have the upgrade label", node.Name))
+				currNodesToUpgrade[node.Name] = node
+			} else {
+				// If the version is outdated but it already has the upgrade label
+				// we do nothing since we're waiting for the node service to finish
+				// working.
+				currNodesUpgrading[node.Name] = node
+			}
+		} else {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the latest version", node.Name))
+			// If the node has the latest version and it still has the upgrade
+			// script label, then it has just finished upgrading and needs the label
+			// removed. If it does not have the label, it's already upgraded and
+			// nothing needs to be done.
+			if _, exists := node.Labels[common.CalicoWindowsUpgradeScriptLabel]; exists {
+				windowsLog.V(1).Info(fmt.Sprintf("Node %v has finished the upgrade", node.Name))
+				currNodesFinishedUpgrade[node.Name] = node
+			}
+		}
+	}
+
+	windowsLog.V(1).Info(fmt.Sprintf("new nodesToUpgrade=%v, nodesUpgrading=%v, nodesFinishedUpgrade=%v", len(currNodesToUpgrade), len(currNodesUpgrading), len(currNodesFinishedUpgrade)))
+
+	// Save copy of previous node upgrade status values.
+	prevNodesToUpgrade := make(map[string]*corev1.Node)
+	prevNodesUpgrading := make(map[string]*corev1.Node)
+	prevNodesFinishedUpgrade := make(map[string]*corev1.Node)
+
+	for k, v := range w.nodesToUpgrade {
```

Comment:
I think we should have locking to protect these shared variables.

Reply:
Fixed in 5e8f813